File: packages/gheim-js/src/openai/_anonymize_only.ts
/** Anonymize-input-only endpoints: embeddings, moderations, images. */
import type OpenAIClass from "openai";
import type { Detector } from "../detectors/base.ts";
import { anonymizeValue, type CallExtras, resolveCallSession } from "./_base.ts";

interface RawCreate {
  create(body: unknown): Promise<unknown>;
}

interface RawImages {
  generate(body: unknown): Promise<unknown>;
  edit(body: unknown): Promise<unknown>;
  createVariation?: (body: unknown) => Promise<unknown>;
}

async function anonymizeFields(
  body: Record<string, unknown> & CallExtras,
  defaultDet: Detector,
  fields: string[],
) {
  const { session, detector, rest } = resolveCallSession(body, defaultDet);
  for (const f of fields) {
    if (f in rest) {
      rest[f] = await anonymizeValue(rest[f], session, detector);
    }
  }
  return { rest, session };
}

export class GheimEmbeddings {
  constructor(
    private readonly clientPromise: Promise<OpenAIClass>,
    private readonly defaultDet: Detector,
  ) {}

  async create(body: Record<string, unknown> & CallExtras): Promise<unknown> {
    const { rest } = await anonymizeFields(body, this.defaultDet, ["input"]);
    const client = await this.clientPromise;
    return (client as unknown as { embeddings: RawCreate }).embeddings.create(rest);
  }
}

export class GheimModerations {
  constructor(
    private readonly clientPromise: Promise<OpenAIClass>,
    private readonly defaultDet: Detector,
  ) {}

  async create(body: Record<string, unknown> & CallExtras): Promise<unknown> {
    const { rest } = await anonymizeFields(body, this.defaultDet, ["input"]);
    const client = await this.clientPromise;
    return (client as unknown as { moderations: RawCreate }).moderations.create(rest);
  }
}

export class GheimImages {
  constructor(
    private readonly clientPromise: Promise<OpenAIClass>,
    private readonly defaultDet: Detector,
  ) {}

  async generate(body: Record<string, unknown> & CallExtras): Promise<unknown> {
    const { rest } = await anonymizeFields(body, this.defaultDet, ["prompt"]);
    const client = await this.clientPromise;
    return (client as unknown as { images: RawImages }).images.generate(rest);
  }

  async edit(body: Record<string, unknown> & CallExtras): Promise<unknown> {
    const { rest } = await anonymizeFields(body, this.defaultDet, ["prompt"]);
    const client = await this.clientPromise;
    return (client as unknown as { images: RawImages }).images.edit(rest);
  }

  async createVariation(body: Record<string, unknown> & CallExtras): Promise<unknown> {
    const client = await this.clientPromise;
    const inner = (client as unknown as { images: RawImages }).images;
    if (!inner.createVariation) {
      throw new Error("openai SDK has no images.createVariation in this version");
    }
    return inner.createVariation(body as unknown);
  }
}

File: packages/gheim-py/src/gheim/openai/_anonymize_only.py
"""Endpoints that send text but receive no text — embeddings, moderations, audio.speech, images."""
from __future__ import annotations

from dataclasses import dataclass
from typing import Annotated, Any

from ..core.session import Session
from ..detectors import Detector
from ._base import anonymize_value, resolve_call_session


def _anonymize_keys(kwargs: dict[str, Any], keys: list[str], sess: Session, det: Detector) -> None:
    for k in keys:
        if k in kwargs:
            kwargs[k] = anonymize_value(kwargs[k], sess, det)


# ---------- Embeddings ----------

@dataclass
class EmbeddingsProxy:
    inner: Annotated[Any, "openai client's embeddings object."]
    detector: Annotated[Detector, "Default detector for this client."]

    def create(self, *args: Any, **kwargs: Any) -> Any:
        sess, detector = resolve_call_session(kwargs, self.detector)
        _anonymize_keys(kwargs, ["input"], sess, detector)
        return self.inner.create(*args, **kwargs)


@dataclass
class AsyncEmbeddingsProxy:
    inner: Annotated[Any, "openai async client's embeddings object."]
    detector: Annotated[Detector, "Default detector for this client."]

    async def create(self, *args: Any, **kwargs: Any) -> Any:
        sess, detector = resolve_call_session(kwargs, self.detector)
        _anonymize_keys(kwargs, ["input"], sess, detector)
        return await self.inner.create(*args, **kwargs)


# ---------- Moderations ----------

@dataclass
class ModerationsProxy:
    inner: Annotated[Any, "openai client's moderations object."]
    detector: Annotated[Detector, "Default detector for this client."]

    def create(self, *args: Any, **kwargs: Any) -> Any:
        sess, detector = resolve_call_session(kwargs, self.detector)
        _anonymize_keys(kwargs, ["input"], sess, detector)
        return self.inner.create(*args, **kwargs)


@dataclass
class AsyncModerationsProxy:
    inner: Annotated[Any, "openai async client's moderations object."]
    detector: Annotated[Detector, "Default detector for this client."]

    async def create(self, *args: Any, **kwargs: Any) -> Any:
        sess, detector = resolve_call_session(kwargs, self.detector)
        _anonymize_keys(kwargs, ["input"], sess, detector)
        return await self.inner.create(*args, **kwargs)


# ---------- Images (generate / edit) ----------

@dataclass
class ImagesProxy:
    inner: Annotated[Any, "openai client's images object."]
    detector: Annotated[Detector, "Default detector for this client."]

    def generate(self, *args: Any, **kwargs: Any) -> Any:
        sess, detector = resolve_call_session(kwargs, self.detector)
        _anonymize_keys(kwargs, ["prompt"], sess, detector)
        return self.inner.generate(*args, **kwargs)

    def edit(self, *args: Any, **kwargs: Any) -> Any:
        sess, detector = resolve_call_session(kwargs, self.detector)
        _anonymize_keys(kwargs, ["prompt"], sess, detector)
        return self.inner.edit(*args, **kwargs)

    # create_variation has no text input — pass through.
    def create_variation(self, *args: Any, **kwargs: Any) -> Any:
        return self.inner.create_variation(*args, **kwargs)

    def __getattr__(self, name: str) -> Any:
        return getattr(self.inner, name)


@dataclass
class AsyncImagesProxy:
    inner: Annotated[Any, "openai async client's images object."]
    detector: Annotated[Detector, "Default detector for this client."]

    async def generate(self, *args: Any, **kwargs: Any) -> Any:
        sess, detector = resolve_call_session(kwargs, self.detector)
        _anonymize_keys(kwargs, ["prompt"], sess, detector)
        return await self.inner.generate(*args, **kwargs)

    async def edit(self, *args: Any, **kwargs: Any) -> Any:
        sess, detector = resolve_call_session(kwargs, self.detector)
        _anonymize_keys(kwargs, ["prompt"], sess, detector)
        return await self.inner.edit(*args, **kwargs)

    async def create_variation(self, *args: Any, **kwargs: Any) -> Any:
        return await self.inner.create_variation(*args, **kwargs)

    def __getattr__(self, name: str) -> Any:
        return getattr(self.inner, name)
